zfs-replicate.sh

```diff
@@ -1,524 +1,538 @@
+        $ZFS snapshot -r ${local_set}@${sname}
+      fi
+    elif [ ${MODE} = PULL ] && [ ${TYPE} = REMOTE ]; then
+      printf "Creating ZFS snapshot %s@%s\n" "${remote_set}" "${sname}"
+      if [ $RECURSE_CHILDREN -ne 1 ]; then
+        printf "RUNNING: %s snapshot %s@%s\n" "${ZFS}" "${remote_set}" "${sname}"
+        ssh $REMOTE_SERVER $ZFS snapshot ${remote_set}@${sname}
+      else
+        printf "RUNNING: %s snapshot -r %s@%s\n" "${ZFS}" "${remote_set}" "${sname}"
+        ssh $REMOTE_SERVER $ZFS snapshot -r ${remote_set}@${sname}
+      fi
+    elif [ ${MODE} = PUSH ] && [ ${TYPE} = LOCAL ]; then
+      printf "Creating ZFS snapshot %s@%s\n" "${local_set}" "${sname}"
+      if [ $RECURSE_CHILDREN -ne 1 ]; then
+        printf "RUNNING: %s snapshot %s@%s\n" "${ZFS}" "${local_set}" "${sname}"
+        $ZFS snapshot ${local_set}@${sname}
+      else
+        printf "RUNNING: %s snapshot -r %s@%s\n" "${ZFS}" "${local_set}" "${sname}"
+        $ZFS snapshot -r ${local_set}@${sname}
+      fi
+    elif [ ${MODE} = PULL ] && [ ${TYPE} = LOCAL ]; then
+      printf "Creating ZFS snapshot %s@%s\n" "${remote_set}" "${sname}"
+      if [ $RECURSE_CHILDREN -ne 1 ]; then
+        printf "RUNNING: %s snapshot %s@%s\n" "${ZFS}" "${remote_set}" "${sname}"
+        ${ZFS} snapshot ${remote_set}@${sname}
+      else
+        printf "RUNNING: %s snapshot -r %s@%s\n" "${ZFS}" "${remote_set}" "${sname}"
+        ${ZFS} snapshot -r ${remote_set}@${sname}
+      fi
+    fi
+
+    ## check return
+    local status=$?
+    if [ $status -ne 0 ]; then
+      exit_clean $status
+    fi
+
+    ## Send incremental snaphot if count is 1 or more, otherwise send full snapshot
+    if [ ${MODE} == PUSH ]; then
+      if [ $scount -ge 1 ]; then
+        if ! do_push ${base_snap} ${local_set}@${sname} ${remote_set}; then
+          do_destroy ${local_set}@${sname}
+          exit_clean 99 ## TODO: use proper status
         fi
-        ## Send incremental snaphot if count is 1 or more, otherwise send full snapshot
-        if [ ${MODE} == PUSH ]; then    
-            if [ $scount -ge 1 ]; then
-                if ! do_push ${base_snap} ${local_set}@${sname} ${remote_set}; then
-		do_destroy ${local_set}@${sname}
-		exit_error
-		fi
-            else
-                if ! do_push "NULL" ${local_set}@${sname} ${remote_set}; then
-		do_destroy ${local_set}@${sname}
-		exit_error
-		fi
-            fi
-        elif [ ${MODE} == PULL ]; then
-            if [ $scount -ge 1 ]; then
-		if ! do_pull ${base_snap} ${remote_set}@${sname} ${local_set}; then
-		do_destroy ${remote_set}@${sname}
-		exit_error
-		fi
-            else
-                if ! do_pull "NULL" ${remote_set}@${sname} ${local_set}; then
-		do_destroy ${remote_set}@${sname}
-		exit_error
-		fi
-            fi
+      else
+        if ! do_push "NULL" ${local_set}@${sname} ${remote_set}; then
+          do_destroy ${local_set}@${sname}
+          exit_clean 99 ## TODO: use proper status
```

Comment:
Why `exit_clean` here? If the replication fails for any reason, it should `exit_error`